client/src/resources/ItemActionController.js

```diff
@@ -0,0 +1,109 @@
+import QConfig from "resources/QConfig.js";
+
+@inject(Router, Notification, I18N, DialogService, QConfig)
+export default class ItemActionController {
+  constructor(router, notification, i18n, dialogService, qConfig) {
+    this.router = router;
+    this.notification = notification;
+    this.i18n = i18n;
+    this.dialogService = dialogService;
+    this.qConfig = qConfig;
+  }
+
+  activate(item) {
+    item.activate();
+  }
+
+  deactivate(item) {
+    item.deactivate();
+  }
+
+  async edit(item) {
+    const editorRoute = `/editor/${item.conf.tool}/${item.conf._id}`;
+
+    // if the item is already active and the last edited and uiBehaviour.askBeforeEditIfActive is true in config, we
+    // ask the user if she really wants to edit this item
+
+    // if the item is not active, go to the editor
+    if (item.conf.active !== true) {
+      this.router.navigate(editorRoute);
+      return;
+    }
+
+    // otherwise we need to load the uiBehaviour Configuration
+    const uiBehaviorConfig = await this.qConfig.get("uiBehavior");
+
+    // if we do not have the config, go to the editor
+    if (!uiBehaviorConfig || !uiBehaviorConfig.askBeforeEditIfActive) {
+      this.router.navigate(editorRoute);
+      return;
+    }
+
+    // otherwise compute the secondsSinceLastEdit
+    const lastEditedDate = new Date(item.conf.updatedDate || Date.now());
+    const now = new Date();
+    const secondsSinceLastEdit =
+      now.getTime() / 1000 - lastEditedDate.getTime() / 1000;
+
+    // if the configured threshold is not reached, go to the editor
+    if (
+      !uiBehaviorConfig.askBeforeEditIfActive.lastEditSecondsThreshold ||
+      secondsSinceLastEdit <
+        uiBehaviorConfig.askBeforeEditIfActive.lastEditSecondsThreshold
+    ) {
+      this.router.navigate(editorRoute);
+      return;
+    }
+
+    // open the dialog to ask the user if she wants to edit
+    const openDialogResult = await this.dialogService.open({
+      viewModel: ConfirmDialog,
+      model: {
+        confirmQuestion: this.i18n.tr("editor.questionReallyEditBecauseActive"),
+        proceedText: this.i18n.tr("editor.confirmProceedToEditor"),
+        cancelText: this.i18n.tr("editor.cancelProceedToEditor")
+      }
+    });
+    const closeResult = await openDialogResult.closeResult;
+
+    // only go to the editor if it was not cancelled
+    if (!closeResult.wasCancelled) {
+      this.router.navigate(editorRoute);
+    }
+  }
+
+  async delete(item, afterDeleteAction) {
+    const openDialogResult = await this.dialogService.open({
+      viewModel: ConfirmDialog,
+      model: {
+        confirmQuestion: this.i18n.tr("item.questionDeleteItem"),
+        proceedText: this.i18n.tr("item.confirmDeleteItem"),
+        cancelText: this.i18n.tr("item.cancelDeleteItem")
+      }
+    });
+    const closeResult = await openDialogResult.closeResult;
+
+    if (!closeResult.wasCancelled) {
+      item.delete().then(() => {
```

Comment:
return the promise of `item.delete` here instead of working with a callback.
in case the user cancels the confirm dialog you could return a rejected promise to handle it correctly wherever this is called.

Reply:
ah yes makes it much nicer -> https://github.com/nzzdev/Q-editor/pull/133/commits/1158ee8e70ecb7f0adf1542b666047acda0cfca7